### DISCUSSION TITLE
Will waveshare ips screen work with esp32 and tft_espi?

### DISCUSSION THREAD

**Discussion by rradomir** [Bodmer/TFT_eSPI#2398] [Compatible displays and setup files]
Hi, for now Im using generic tft screen 320x240 with ILI9341 and XPT2046 on ESP32 devkit board.
I want to upgrade to this screen: https://www.waveshare.com/wiki/Pico-ResTouch-LCD-3.5 (_They even refers to yours library  :)_ )
Will it work? it was designed to work with raspberry pi pico, but SPI is SPI right? It should work...? ESP32 is even faster than pico..?
Driver ILI9488 is listed in readme. Thanks in advance!

**Comment by Bodmer** [comment]
I would not choose that screen unless a RP2040 Pico is used. Also that screen has a bespoke serial to parallel interface that has been used on the Linux Raspberry Pi boards. It uses a shift register and typically does not work above 20MHz SPI clock rate.

See also #2357 where a user was successful using an RP2040 Pico with that board but then could not get it working with an Pico footprint ESP32 S3.

**Comment by rradomir** [comment]
Well. 
As you predicted, i had problems with getting esp32 work with this screen. However I bought Pico and it actually works with this screen with setup files from their page (example file Setup60_RP2040_ILI9341 is slightly different) and i had inverted colors (but TFT_INVERSION_ON did the job).
Backlight can be PWM driven - nice feature. It isn't described.
Sadly, SPI speed, or PICO speed cannot fill screen at once. Even when I set enormous speed in SPI_FREQUENCY. 
In theory, when RPI_DISPLAY_TYPE is defined, SPI speed is max 20Mhz, but I clearly see difference, when I set SPI_FREQUENCY to higher value. But still too slow.
If display_type isn't defined, screen don't work at all. I don't know it could be possible by esp32 also, maybe this is just SPI limitation.

If someone has clue how to marriage esp32 with this screen it would be great :) maybe is there a "hack"? Im rather beginner.


I **asked waveshare support** same question, and they answered, that it should be possible. They pointed this https://www.waveshare.com/wiki/ESP32-S2-Pico#Examples_2 as it should be help me to do it.

**Reply by Bodmer** [reply]
See this discussion:
https://github.com/earlephilhower/arduino-pico/discussions/415

In summary:
1. The maximum SPI clock rate is 62.5MHz when the processor clock is 125MHz.
2. The maximum SPI rate drops to 24MHz maximum when the processor is clocked at 133MHz unless the SPI source clock is changed (see link above).

**Reply by rradomir** [reply]
Thank you. 
It's faster now. When frequency is too fast generates glitches, wrong colors or mirrored image (but clears screen instant:). Trial and error method shows that speed 67MHz works well. Even if you said, that 62.5 is max, there is slightly difference. I recorded video of simple filling screen, so I have proof ;) 
Great library tho! Funny examples and brilliant, useful ideas, I like them. Now I need to run built-in SD card :D

**Reply by PJ5168** [reply]
Dear Sir,
SD is another issue for RP2040... slow as well.
Still trying to make the 4-bit SDIO to work.
If you know how to make ESP32 works on RPi type ILI9488 LCD, please let me know.
Appreciate!

PJ.

**Reply by PJ5168** [reply]
Dear,
I made the example code works!
The ESP32-S3 now display the screen on the Waveshare 3.5" RPI type ILI9488 LCD.
This proof that the HW is working.
The next step is to find out the difference between TFT_eSPI and the example code.
